integration-testing/src/infra/subnet.rs

```diff
@@ -0,0 +1,562 @@
+        node.join_subnet().await?;
+        log::info!(
+            "validator: {:?} joined subnet: {:}",
+            node.validator.net_addr,
+            node.id
+        );
+
+        sleep(std::time::Duration::from_secs(5));
+
+        node.spawn_validator()?;
+        log::info!("validator: {:?} spawned", node.validator.net_addr);
+    }
+
+    print_toml_config(topology, &nodes).await?;
+
+    Ok(())
+}
+
+async fn print_toml_config(topology: &SubnetTopology, nodes: &[SubnetNode]) -> Result<()> {
+    println!("\n\n========== SUBNET TOML CONFIG ============ \n\n");
+
+    let accounts = nodes
+        .iter()
+        .map(|n| n.wallet_address.clone().unwrap())
+        .collect::<Vec<_>>()
+        .join("\",\"");
+
+    println!("accounts = [{accounts:?}]");
+
+    let mut admin_token = nodes[0].create_admin_token().await?;
+    trim_newline(&mut admin_token);
+    println!("auth_token = \"{admin_token:}\"");
+
+    println!(
+        "jsonrpc_api_http = \"http://127.0.0.1:{:}/rpc/v1\"",
+        nodes[0].node.tcp_port
+    );
+
+    println!("id = \"{:}\"", topology.id.as_ref().unwrap());
+    println!("gateway_addr = \"t064\"");
+    println!("network_name = \"{:}\"", topology.name);
+
+    Ok(())
+}
+
+fn node_from_topology(topology: &SubnetTopology) -> SubnetNode {
+    SubnetNode::new(
+        topology.id.clone().unwrap(),
+        topology.ipc_root_folder.clone(),
+        topology.next_port(),
+        topology.next_port(),
+        topology.next_port(),
+        topology.next_port(),
+        topology.eudico_binary_path.clone(),
+        topology.ipc_agent_url(),
+    )
+}
+
+/// Spawn the first node, then subsequent node will connect to this node.
+fn spawn_first_node(topology: &SubnetTopology) -> anyhow::Result<SubnetNode> {
+    let mut node = node_from_topology(topology);
+    node.gen_genesis()?;
+    node.spawn_node()?;
+
+    util::create_wallet(&mut node)?;
+    node.config_default_wallet()?;
+    Ok(node)
+}
+
+fn spawn_other_nodes(
+    topology: &SubnetTopology,
+    first: &SubnetNode,
+) -> anyhow::Result<Vec<SubnetNode>> {
+    let mut nodes = vec![];
+    for _ in 1..topology.number_of_nodes {
+        let mut node = node_from_topology(topology);
+
+        node.spawn_node()?;
+
+        util::create_wallet(&mut node)?;
+
+        node.config_default_wallet()?;
+
+        nodes.push(node);
+    }
+
+    let addrs = loop {
+        match first.network_addresses() {
+            Ok(s) => {
+                break s;
+            }
+            Err(e) => {
+                log::warn!("first node not up, wait: {e:}");
+                sleep(std::time::Duration::from_secs(5));
+            }
+        }
+    };
+
+    let mut first_node_addr = util::tcp_address(addrs)?;
+    util::trim_newline(&mut first_node_addr);
+
+    log::info!("first node net addr: {:?}", first_node_addr);
+
+    for node in &nodes {
+        node.connect_peer(&first_node_addr)?;
+    }
+
+    Ok(nodes)
+}
+
+pub struct SubnetNode {
```

Comment:
Should we move all node-specific code to a `node.rs` file and leave here subnet info or they are too related?

Reply:
Also, would you mind commenting a bit the structs and functions of this file?